### PR TITLE
[4.0] Adaptive Accessibility Menu

### PR DESCRIPTION
**Pull request by brianteeman**
Read more about the plugin here https://ranbuch.github.io/accessibility/

This was previously at #24112 but there was a bug upstream which has now been fixed

Feature configuration options have not been added at this point - not sure if we should.

This can be configured to be site, administrator or both

To test you will need to
1. apply the patch
2. composer install
3. npm i
4. Discover the plugin
5. Enable the plugin

Supports RTL

image

image

### Previous Comment 
@zwiastunsw 
> * Such a plugin is useful and may improve the accessibility of the site, provided that it works well. Brian has chosen a very good script that doesn't cause the errors that widgets of this type cause.
> * The plugin does not solve the problems of people with disabilities. People with disabilities use specialised tools (screen readers, screen magnifiers, braille embossers… etc.). But it improves usability for the elderly, people with visual impairments, people with cognitive problems.
> * E.g. a blind person will use a screen reader and not the option to change text to speech. But this is a very good solution for the elderly, for people who have difficulty concentrating, and also for those who cannot read at the moment.
> * Text enlargement is very useful for many seniors and people who see less well. Such people often do not know about the possibilities of the browser and like to use such tools.
> * The contrast version will not replace the correct contrasts on the website. But it can be helpful for people who have weaker eyesight or for people who browse the site, for example in a sunny place.
> * WCAG does not require the links to be underlined (they must be highlighted). But underlined links can improve accessibility for visually impaired people or for people who have difficulty concentrating.

### CODE REVIEW

**Comment by brianteeman** [comment]
@fedik I fixed the bits I wanted help on. Well it works but maybe it could be better code.

**Comment by Fedik** [comment]
it looks good, I only would use WebAsset :wink: 
I can make a pull, but a bit later

**Comment by Fedik** [comment]
I have tested a bit, and noticed the font size scaling works "not equal" for all blocks,
I think it is due to how font size coded in the template style

**Comment by brianteeman** [comment]
> TEXT SIZE MANIPULATION APPROACHES
If text increase / decrease isn't working for your size your probablly not using responsive font size units (sutch as em, rem etc.).

> In that case you can initialize the accessibility tool like this:

> new Accessibility({textPixelMode: true})

**Comment by Fedik** [comment]
> In that case you can initialize the accessibility tool like this: 
new Accessibility({textPixelMode: true})

Yeah, but it have a huge performance drawback, I just tried.
It is better if someone review hardcoded font sizes in the template, after the plugin will be merged

**Comment by brianteeman** [comment]
> It is better if someone review hardcoded font sizes in the template, after the plugin will be merged

I agree

**Comment by brianteeman** [comment]
thanks for the web asset stuff - I will test and merge that soon

**Comment by Fedik** [comment]
hm, no, sorry for misleading, the fonts is not hardcoded,

I see the plugin manipulate font size of `<body>`, that not always correct, 
 would be better if it can do with `<html>`
All fonts in template are in `rem` that is allow you to change font size in browser settings, that how browser Accessibility works (see font size in browser settings)

**Comment by brianteeman** [comment]
Can you tell me where you are seeing fonts not resizing

**Comment by Fedik** [comment]
I tried administrator/index.php?option=com_plugins&view=plugins
the menu "resizes", but the content in table and in filters are not, 
also if open any plugin to edit

but resizing work, if I change font size in browser setting

**Comment by brianteeman** [comment]
bugger - confirmed

**Comment by brianteeman** [comment]
Now uses web asset

**Comment by brianteeman** [comment]
@Fedik the good news is that its nothing to do with the script and it is some sort of bug with the current atum template. It works perfectly with the template @angieradtke is working on

**Comment by Fedik** [comment]
>  bug with the current atum template

There no bug in template, the script use incorrect approach.
Check section "Base font size and accessibility"
https://betterwebtype.com/articles/2019/06/16/5-keys-to-accessible-web-typography/

But if the template will change all font size to `em` then it will work as is, of course. That also fine.

**Comment by Fedik** [comment]
btw, I have open an issue about font-size bug, in their repository.
Would be nice if they can fix it on their side.

**Comment by Fedik** [comment]
he fixed the font issue, can you try update script here to 3.0.9

**Comment by brianteeman** [comment]
awesome - will do that this afternoon

**Comment by brianteeman** [comment]
updated the version

**Comment by Fedik** [comment]
I have tested this item :white_check_mark: successfully on e1eb82ee2268bab13bdb410816f9b4ce07d76a45<br /><br />now it works much better<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/27940">issues.joomla.org/tracker/joomla-cms/27940</a>.</sub>

**Comment by alikon** [comment]
I have tested this item :white_check_mark: successfully on e1eb82ee2268bab13bdb410816f9b4ce07d76a45<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/27940">issues.joomla.org/tracker/joomla-cms/27940</a>.</sub>

**Comment by alikon** [comment]
RTC<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/27940">issues.joomla.org/tracker/joomla-cms/27940</a>.</sub>

**Comment by alikon** [comment]
btw a bit ugly when you invert colours ...
is this configurable ?

File: plugins/system/accessibility/accessibility.php [hunk 1]
```diff
@@ -0,0 +1,105 @@
...
+ * @subpackage  System.accessibility
+ *
+ * @copyright   Copyright (C) 2005 - 2019 Open Source Matters, Inc. All rights reserved.
+ * @license     GNU General Public License version 2 or later; see LICENSE.txt
+ */
+
+defined('_JEXEC') or die;
+
+use Joomla\CMS\Factory;
+use Joomla\CMS\HTML\HTMLHelper;
```
Comment (Quy) [inline plugins/system/accessibility/accessibility.php]:
```suggestion
```

**Comment by brianteeman** [comment]
1. please remove rtc as there is no sql yet
2. invert colours is supposed to look like that. its not for your eyesight

**Comment by brianteeman** [comment]
sql files now added and this PR is complete

**Comment by brianteeman** [comment]
conflicts resolved

**Comment by alikon** [comment]
I have tested this item :white_check_mark: successfully on c4bf48ccaa26dd9aa8b5df8bd88bbe2f5bed8c4e<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/27940">issues.joomla.org/tracker/joomla-cms/27940</a>.</sub>

**Comment by jwaisner** [comment]
I have tested this item :white_check_mark: successfully on c4bf48ccaa26dd9aa8b5df8bd88bbe2f5bed8c4e<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/27940">issues.joomla.org/tracker/joomla-cms/27940</a>.</sub>

**Comment by jwaisner** [comment]
RTC<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/27940">issues.joomla.org/tracker/joomla-cms/27940</a>.</sub>

**Comment by brianteeman** [comment]
remaining drone error is unrelated to this pr and fixed with #28293

**Comment by wilsonge** [comment]
Can you merge through 4.0-dev please so I can see the green tick of goodness please :)

File: plugins/system/accessibility/accessibility.php [hunk 2]
```diff
@@ -0,0 +1,104 @@
...
+		$this->loadLanguage();
+
+		// Determine if it is an LTR or RTL language
+		$direction = Factory::getLanguage()->isRTL() ? 'left' : 'right';
+
+		/**
+		* Add strings for translations in Javascript.
+		* Reference  https://ranbuch.github.io/accessibility/
+		*/
+			$document->addScriptOptions(
```
Comment (wilsonge) [inline plugins/system/accessibility/accessibility.php]: I'm boring but looks like you've got an extra indent here

**Comment by wilsonge** [comment]
Other than a boring piece of codestyle this looks good to me

File: plugins/system/accessibility/accessibility.xml [hunk 3]
```diff
@@ -0,0 +1,36 @@
+<?xml version="1.0" encoding="utf-8"?>
+<extension version="4.0" type="plugin" group="system" method="upgrade">
```
Comment (Quy) [inline plugins/system/accessibility/accessibility.xml]:
```suggestion
<extension type="plugin" group="system" method="upgrade">
```

Comment (Quy) [reply plugins/system/accessibility/accessibility.xml]: See #27969

**Comment by brianteeman** [comment]
@wilsonge all done - the new drone failures are the usual pos test failing

**Comment by wilsonge** [comment]
ta. i'll keep kicking it until it goes green

File: libraries/src/Extension/ExtensionHelper.php [hunk 4]
```diff
@@ -263,6 +263,7 @@ class ExtensionHelper
 		array('plugin', 'multilang', 'sampledata', 0),
 
 		// Core plugin extensions - system
+		array('plugin', 'accessibility', system, 0),
```
Comment (wilsonge) [inline libraries/src/Extension/ExtensionHelper.php]:
```suggestion
		array('plugin', 'accessibility', 'system', 0),
```

Actually tests are correctly failing.  https://ci.joomla.org:444/artifacts/joomla/joomla-cms/4.0-dev/27940/system-tests/30108/ContentCategoryListCest.Category.mysql.fail.png

**Comment by brianteeman** [comment]
all green now

**Comment by wilsonge** [comment]
And all merged - thanks!

**Comment by brianteeman** [comment]
Yay. Thanks

**Comment by Phil123321** [comment]
Do you want to have a great time? The <a href="https://www.1depositcasinonz.com/" rel="nofollow">1 dollar deposit casino New Zealand</a> will help you with this<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/27940">issues.joomla.org/tracker/joomla-cms/27940</a>.</sub>

**Pull request merged**